convert-hf-to-powerinfer-gguf.py

```diff
@@ -0,0 +1,601 @@
+#!/usr/bin/env python3
+
+from __future__ import annotations
+from abc import ABC, abstractmethod
+
+import argparse
+import contextlib
+import json
+import os
+import re
+import struct
+import sys
+from enum import IntEnum
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, ContextManager, Iterator, Optional, cast
+
+import numpy as np
+import torch
+import torch.nn as tnn
+
+if TYPE_CHECKING:
+    from torch import Tensor
+
+if "NO_LOCAL_GGUF" not in os.environ:
+    sys.path.insert(1, str(Path(__file__).parent / "gguf-py"))
+import gguf
+
+
+###### MODEL DEFINITIONS ######
+
+
+class SentencePieceTokenTypes(IntEnum):
+    NORMAL = 1
+    UNKNOWN = 2
+    CONTROL = 3
+    USER_DEFINED = 4
+    UNUSED = 5
+    BYTE = 6
+
+
+class ReluMLP(tnn.Module):
+    def __init__(self, input_dim: int, hidden_dim: int, output_dim: int):
+        super(ReluMLP, self).__init__()
+        self.fc1 = tnn.Linear(input_dim, hidden_dim, bias=False)
+        self.relu = tnn.ReLU()
+        self.fc2 = tnn.Linear(hidden_dim, output_dim, bias=False)
+
+    def forward(self, x):
+        x = self.fc1(x)
+        x = self.relu(x)
+        x = self.fc2(x)
+        return x
+
+    @staticmethod
+    def from_file(model_file: Path):
+        model = torch.load(model_file, map_location="cpu")
+        hidden_size, input_size = model.get("fc1.weight").shape
+        output_size, _ = model.get("fc2.weight").shape
+        mlp = ReluMLP(input_size, hidden_size, output_size)
+        mlp.load_state_dict(model)
+        return mlp
+
+
+class Model(ABC):
```

Comment:
GGUF converter base class. Possibly reused for loading LLAMA and Falcon model files.